### DISCUSSION TITLE
Could not load file or assembly 'CefSharp.Core

### DISCUSSION THREAD

**Discussion by A6932401** [cefsharp/CefSharp#3373] [General]
Dear Team,
I used CefSharp 85.3.130 in VS 2015 existing Windows Application. But during run the application "Could not load file or assembly 'CefSharp.Core, Version=86.0.241.0, Culture=neutral, PublicKeyToken=40c4b6fc221f4138' or one of its dependencies. The specified module could not be found." this error came. 
I tried all possible way.

One more point its worked for when i create new project and add cefsharp.

So, Please give me some solution.

Thanks,
ArunRaja

**Comment by amaitland** [comment]
> I used CefSharp 85.3.130 in VS 2015 existing Windows Application. But during run the application "Could not load file or assembly 'CefSharp.Core, Version=86.0.241.0,

Which version are you actually using? Your version numbers don't match. 

When are you getting the error on your development pc or when deploy your application? 

> One more point its worked for when i create new project and add cefsharp.

What do you mean by this exactly? 

> I tried all possible way.

What exactly have you tried? You need to provide more information.

**Reply by A6932401** [reply]
Point 1: Installed CefSharp.Winform 86.0.241 in VS 2017 Windows Application. 
Point 2: Yes error getting in My PC. when  Rebuild the project then mentioned error came. So, I can't publish the Application.
Ponit 3: Its working Just i create new Project and add Browser Configuration. Next step I try to add this configuration Existing My Click Once windows application and run the application but mentioned error came.
Point 4: Please refer attachment. 
[cefsharp.txt](https://github.com/cefsharp/CefSharp/files/5927042/cefsharp.txt)

Hope you Understand my Issue.
Please Help me ASAP.

Thanks,
Arun Raja

**Reply by amaitland** [reply]
Upgrade to version 87 and add the property to your project file as described in https://github.com/cefsharp/CefSharp/issues/3319#issuecomment-757106286

**Reply by A6932401** [reply]
Version 87 is Support for .net framework 4.6.1?

Bcz, After Update the Package " Could not load file or assembly 'CefSharp.Core.Runtime.dll' or one of its dependencies. The specified module could not be found." Error came.

**Reply by amaitland** [reply]
> Version 87 is Support for .net framework 4.6.1?

Yes

> Bcz, After Update the Package " Could not load file or assembly 'CefSharp.Core.Runtime.dll' or one of its dependencies. The specified module could not be found." Error came.

`CefSharp.Core.dll` is effectively now `CefSharp.Core.Runtime.dll`.

You need to remove all the changes you made previously for supporting `ClickOnce` and just apply the one I linked above.

**Reply by A6932401** [reply]
I applied all you mentioned Changes also Updated CefSharp Version 87. Now i got "CefSettings.BrowserSubprocessPath now requires an absolute path, the path provided is non-absolute. You can use System.IO.Path.GetFullPath to obtain an absolute path. Current value:x86\CefSharp.BrowserSubprocess.exe"  this Error came.

After configuration first time Run the project attached screen open for ask path. I click cancel button.
Next time screen not come.
Is Correct? 
![cefsharp](https://user-images.githubusercontent.com/73706751/107021922-be60c600-67ca-11eb-820a-1743442035ce.PNG)

**Reply by amaitland** [reply]
As the exception says, BrowserSubprocessPath needs to be an absolute path. 

You can use https://docs.microsoft.com/en-us/dotnet/api/system.io.path.getfullpath?view=net-5.0#System_IO_Path_GetFullPath_System_String_ to get an absolute path from a relative one.

**Reply by A6932401** [reply]
I tried u mentioned solution but "CefSettings.BrowserSubprocessPath not found." error came.
Could please give some more detailed solution?
i attached program.cs screen shot.  Could please check it?


![cefsharp1](https://user-images.githubusercontent.com/73706751/107107777-4f7e7e00-6859-11eb-9d2c-f0aac4cbc785.PNG)

**Reply by amaitland** [reply]
Just comment out the line, the default will work in 95% of cases.

If you still have a problem the provide an absolute path to the browsersubprocess executable.

**Reply by A6932401** [reply]
Its worked...  Thanks You lot...

**Comment by pvg8v6g** [comment]
I am also getting a file not found.

`System.Windows.Markup.XamlParseException: The invocation of the constructor on type 'CefSharp.Wpf.ChromiumWebBrowser' that matches the specified binding constraints threw an exception. ---> System.IO.FileNotFoundException: Could not load file or assembly 'CefSharp.Core.Runtime.dll' or one of its dependencies. The specified module could not be found.`

This error only occurs on the published version. It never occurs in debug or release mode while running through VS.

Really not sure what the issue is and I have every c++ redistributable installed on planet earth plus all the files in my bin that are listed [here](https://github.com/cefsharp/CefSharp/wiki/Output-files-description-table-(Redistribution))

HELP!!

**Reply by amaitland** [reply]
Add the property to your project file as described in https://github.com/cefsharp/CefSharp/issues/3319#issuecomment-757106286

**Reply by pvg8v6g** [reply]
Unfortunately that did not work.

The weird thing is: most of these file not found bug reports are about the CefSharp.Core.dll missing but mine is the CefSharp.Core.Runtime.dll

I even used a dependency dump on that file to see what files it's referencing and I have every one of them.

I'm really at a loss here.

**Reply by pvg8v6g** [reply]
> 
> 
> Add the property to your project file as described in [#3319 (comment)](https://github.com/cefsharp/CefSharp/issues/3319#issuecomment-757106286)

BTW I LOVE this browser. It does exactly what we need it to do and runs perfectly in visual studio but then as soon as you publish it throws an error.

I'm available to give you any info you need for us to get this running.

**Reply by amaitland** [reply]
> The weird thing is: most of these file not found bug reports are about the CefSharp.Core.dll missing but mine is the CefSharp.Core.Runtime.dll

This is expected as of version 87 see #3319

> but then as soon as you publish it throws an error.

Please provide details about how you are publishing. 

- Visual studio version
- project type (SDK style vs older non SDK style https://docs.microsoft.com/en-us/nuget/resources/check-project-format)
- What publish settings are you using 
- .Net version you are targeting. 

Check your publish folder to determine if all the resources are present

**Reply by pvg8v6g** [reply]
Yeah thanks for the info on the core vs runtime file.

- VS = 2019
- Project Type = WPF project selected in VS. Doesn't look like SDK by the project file though.
![2021-02-05 10_39_42-C__Users_pvaughn_Documents_C Projects_PublishTestApp_PublishTestApp_PublishTestA](https://user-images.githubusercontent.com/19214964/107062056-7deb5300-679e-11eb-852d-f1ea8b444723.png)

- Publish Settings = brand new project for trying to get this to publish. tell me what i need in the publish settings and i'll put them there
- .Net = 4.7.2

I have just a brand new project to get this working. I'll migrate it to my actual real project once i get it to publish. The new project needs to be set up the way it needs to to work.

**Reply by amaitland** [reply]
Where did you add the CefSharpBuildAction property to your project file?

**Reply by pvg8v6g** [reply]
I do not have that property. Is it required? (image taken from [here](https://github.com/cefsharp/CefSharp/blob/master/NuGet/CefSharp.Common.targets))

![image](https://user-images.githubusercontent.com/19214964/107082451-cebc7500-67b9-11eb-96c9-c978a2a9fcd4.png)

Looks to me like there are quite a few options that could go in there.

**Reply by amaitland** [reply]
You said above that you've already added it, was that incorrect? 

https://github.com/cefsharp/CefSharp/discussions/3373#discussioncomment-339090

**Reply by pvg8v6g** [reply]
Oh my bad. I didn't add that.... I quoted that reply for some reason and then stated something completely different in the response. That's my fault.

Regardless, I need to know what to actually put in there.

**Reply by amaitland** [reply]
> Regardless, I need to know what to actually put in there.

 https://github.com/cefsharp/CefSharp/issues/3319#issuecomment-757106286

**Reply by pvg8v6g** [reply]
https://github.com/pvg8v6g/C_Repository/tree/main/CefSharpTest

Here's my project. You'll have to restore the nuget packages.

I'm sorry I've just been working on a project I need this for for over 3 weeks now. The last piece of the puzzle is a working chromium browser.

**Reply by amaitland** [reply]
Does your app display the browser from within visual studio? 

If yes then have you compared the bin folder to that of the folder where you installed the published version? Also compare the exe config, make sure the dependentAssembly entries exist in your published version. 

I haven't tried publishing with an AnyCPU targetplatform. I don't use clickonce, it's very difficult to debug when there's a problem.

**Reply by amaitland** [reply]
> I haven't tried publishing with an AnyCPU targetplatform. I don't use clickonce, it's very difficult to debug when there's a problem.

I've tried targeting `AnyCPU` and there is a problem which I've detailed at https://github.com/cefsharp/CefSharp/issues/3319#issuecomment-777949648

If you target `x86/x64` it will work. As yet I have no workaround for targeting `AnyCPU`
